### DISCUSSION TITLE
Error estimation in `model.expect_data()`

### DISCUSSION THREAD

**Discussion by BlaiseDelaney** [scikit-hep/pyhf#1509] [Q&A] [answered]
Dear developers,

I wondered if there were a way to return the uncertainties through pyhf when calling `model.expected_data()`. Supposing I have a complex model with additive and multiplicative modifiers (including `normsys` and `staterror`) and the full set of parameters obtained from the likelihood minimisation, how do I propagate the per-parameter uncertainty to the overall model?

Ideally, I would like to understand is whether there exists a method `model.expected_data_ufloat()` which can accept both the central value and the uncertainty of each parameter and returns two tensors. The first is what is presently available through `model.expected_data()` and the second would be the tensor of corresponding uncertainties, obtained through proper error propagation in pyhf.

I have attempted to do this by hand, but I am quite unsure on how to deal with the less trivial modifiers, such as `normsys`.

Any help would be much appreciated.

Thanks!

Blaise

**Comment by alexander-held** [comment, marked as answer]
Hi, there are two ways of doing this: bootstrapping and error propagation. `pyhf` itself does not include a method for this at the moment.

For bootstrapping, see https://github.com/scikit-hep/pyhf/discussions/1187#discussioncomment-915784 and https://github.com/scikit-hep/pyhf/issues/1359#issuecomment-867638285 for an example. You can sample from a multivariate Gaussian and evaluate the model prediction for all samples. In principle you could sample from the full likelihood directly to avoid the Gaussian approximation as well.

Error propagation can for example be done via `iminuit.util.propagate` as described in [this tutorial](https://nbviewer.jupyter.org/github/scikit-hep/iminuit/blob/master/tutorial/how_to_draw_error_bands.ipynb). Another implementation is provided via the `cabinetry` library in [`cabinetry.model_utils.calculate_stdev`](https://github.com/alexander-held/cabinetry/blob/17eb67cdd37714bdd69606e4f440010c6d91a6b5/src/cabinetry/model_utils.py#L176-L285). See this [example notebook](https://nbviewer.jupyter.org/github/cabinetry/cabinetry-tutorials/blob/master/example.ipynb) showing the use of `cabinetry`, and in particular the `cabinetry.visualize.data_MC` call that creates pre- and post-fit model visualization and yield tables with related uncertainties. For the error propagation it is important to also have the parameter correlations available, as correlations typically significantly change the post-fit uncertainty.

**Reply by BlaiseDelaney** [reply]
Thanks @alexander-held! I'll take some time to review your suggestions and understand them. This was very helpful.
